
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Using stitchr &#8212; stitchr  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="stitchr input data" href="inputdata.html" />
    <link rel="prev" title="Getting started" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="using-stitchr">
<span id="usage-label"></span><h1>Using <code class="docutils literal notranslate"><span class="pre">stitchr</span></code><a class="headerlink" href="#using-stitchr" title="Permalink to this heading">¶</a></h1>
<p>This script can take either amino acid or nucleotide sequences for the CDR3 junction. However when submitting an amino acid CDR3 sequence, <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> will in most cases <strong>not</strong> produce the actual recombined sequences that encoded the original TCRs, apart from a few edge cases (such as particularly germ-like like alpha chain rearrangements). In these cases, <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> recreates an equivalent full length DNA sequence that will encode the same protein sequence. It aims to produce a sequence as close to germline as possible, so all CDR3 residues that <em>can</em> be germline encoded by the V and J genes are. Non-templated residues in the CDR3 (or those templated by the D gene, which is treated as non-templated for the purpose of stitching) are chosen from taking the most commonly used codon per residue.</p>
<p>When provided with a nucleotide CDR3 sequence <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> can simply line up the edges of the V and the J and pop it in. (The exception that might still produce slightly different nucleotide sequences is when during non-templated deletion and addition a long stretch of V or J gene nucleotides were removed and then a different sequence coincidentally encoding the same amino acids was introduced.)</p>
<p>Care must be taken to ensure that the correct TCR informaton is input, ensuring that:</p>
<ul>
<li><p><strong>You’re using proper IMGT gene nomenclature</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Older/deprecated gene names will not work</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>You have the correct and full CDR3 junction sequence</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="http://www.imgt.org/FAQ/#question39">I.e. running inclusively from the conserved cysteine to the conserved phenylalanine (or rarely, tryptophan) residues</a></p></li>
<li><p>Either as amino acid or DNA sequences</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>You are using the right alleles for the TCR genes in question</strong> if known</p>
<blockquote>
<div><ul class="simple">
<li><p>I.e. the bit after the asterisk in the gene name</p></li>
<li><p>There are many known non-synonymous polymorphisms (and undoubtedly many more unknown ones) which could be impacting on antigen recognition, surface expression, and other aspects of TCR biology</p></li>
<li><p>For best results, try to get long read TCR sequence data and process it through V/J/CDR3 annotation software which is capable of a) providing allele-level resolution, and b) can take up-to-date germline reference files.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>The script produces a TCR from the information given, trying to provide warnings or errors if it detects an improbable or implausible combination, yet it’s possible that the script might produce output that <em>looks</em> OK yet which does not reproduce a coding sequence for the intended TCR.</p>
<section id="gene-allele-default-behaviour">
<h2>Gene/allele default behaviour<a class="headerlink" href="#gene-allele-default-behaviour" title="Permalink to this heading">¶</a></h2>
<p>We recommend to explicitly state the complete allele name, where possible. However if you request an allele for which there isn’t complete sequence data, the script will attempt to default to the prototypical allele (*01) of that gene, or a preferred default allele if the <code class="docutils literal notranslate"><span class="pre">-p</span></code> flag is set (see below). Similarly it will attempt to use the correct leader sequences (L-PART1+L-PART2) for the specified allele, but if it can’t find one it’ll default back to the prototype’s. In both cases, if it cannot find sequence for that allele then it will attempt to use an alternative allele for which data exists in the reference (see <a class="reference external" href="https://github.com/JamieHeather/stitchr/issues/25#issuecomment-1146626463">this comment for more details</a>). Note that IMGT-provided gene sequences which are ‘partial’ at either end of their sequence are discounted entirely, as full length sequences are needed. If the script is needed to stitch TCRs that make use of genes that are partial at their recombination-distal ends then you can modify the FASTA header for these entries in the installed Data directory.</p>
<p>For human and mouse TCRs, the script will use the TRBC gene located in the same cluster as the J gene (i.e. TRBJ1-1 through TRBJ1-6 will get TRBC1, while TRBJ2-1 through TRBJ2-7 will get TRBC2). This can be overriden (see optional arguments). Unfortunately we are not experts in TCR loci architecture of all species, so we haven’t hard-wired any other constant region assumptions, so for all other species you’ll need to explicitly state which constant region you want used.</p>
</section>
<section id="translation-sequences">
<h2>Translation sequences<a class="headerlink" href="#translation-sequences" title="Permalink to this heading">¶</a></h2>
<p>By default <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> does not include stop codons at the end of the coding sequence; if desired, this must be specified using the 3’ flag (<code class="docutils literal notranslate"><span class="pre">-3p</span></code>), i.e. <code class="docutils literal notranslate"><span class="pre">-3p</span> <span class="pre">TAA</span></code>, <code class="docutils literal notranslate"><span class="pre">-3p</span> <span class="pre">TAG</span></code>, or <code class="docutils literal notranslate"><span class="pre">-3p</span> <span class="pre">TGA</span></code>. Similarly, no sequence is included before that of the IMGT-recorded L1 leader sequence. If desired, this can be added using the 5’ flag (<code class="docutils literal notranslate"><span class="pre">-5p</span></code>), e.g. to add the pre-start codon section of an optimal Kozak sequence: <code class="docutils literal notranslate"><span class="pre">-5p</span> <span class="pre">GCCGCCACC</span></code>. Note that translated sequence in the output is the <em>whole</em> stitched sequence, including any added 5’/3’ sequences: addition of 5’ sequences may cause the introduction of underscores (’_’) to appear in the translated output, representing incomplete codons that could not be translated. Also note that the translated sequence of an individual chain may differ from the corresponding section of a linked heterodimer for this reason, depending on the length/frame of the 5’ rearrangement.</p>
</section>
<section id="seamless-mode">
<h2>Seamless mode<a class="headerlink" href="#seamless-mode" title="Permalink to this heading">¶</a></h2>
<p>If users care about accurately replicating the exact nucleotide sequence of specific V(D)J rearrangements, and they have additional nucleotide sequences beyond the edges of the CDR3 junction, they can make use of the optional <code class="docutils literal notranslate"><span class="pre">-sl</span></code> ‘seamless’ flag to stitch together the complete recombined sequence as faithfully as possible.</p>
<p>E.g. instead of these first two options:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>stitchr<span class="w"> </span>-v<span class="w"> </span>TRBV7-6<span class="w"> </span>-j<span class="w"> </span>TRBJ1-4<span class="w"> </span>-cdr3<span class="w"> </span>CASSSGQGLGEKLFF
stitchr<span class="w"> </span>-v<span class="w"> </span>TRBV7-6<span class="w"> </span>-j<span class="w"> </span>TRBJ1-4<span class="w"> </span>-cdr3<span class="w"> </span>TGTGCCAGCAGTTCCGGACAGGGCTTGGGAGAAAAACTGTTTTTT
</pre></div>
</div>
<p>… you would run (NB non-CDR3 nucleotides shown in lower case for display purposes):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>stitchr<span class="w"> </span>-sl<span class="w"> </span>-v<span class="w"> </span>TRBV7-6<span class="w"> </span>-j<span class="w"> </span>TRBJ1-4<span class="w"> </span>-cdr3<span class="w"> </span>catgtatcgcTGTGCCAGCAGTTCCGGACAGGGCTTGGGAGAAAAACTGTTTTTTggcagtggaa
</pre></div>
</div>
<p>In this example aligning the results shows that the second serine in the CDR3 was actually encoded by ‘AGT’ in the rearrangement: the ‘AGC’ codon present in the germline gene must have been deleted and this alternative ‘S’ codon added or completed by Tdt. Thus while all options should produce the same amino acid sequence, the seamless option allows for truer generation of the sequence as was present in the clonotype. Note that the seamless option adds significantly to the time it takes to run <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> (which only really matters when running it on high-throughput datasets using <code class="docutils literal notranslate"><span class="pre">thimble</span></code>).</p>
<p>In order to best use the seamless option, please ensure that:</p>
<ul>
<li><p>You have sufficient nucleotide context on either side of the CDR3 (especially the V) - ideally 20-30 nucleotides.</p></li>
<li><p>Do not include any leader or constant region nucleotides - this may involve trimming nucleotide sequences.</p></li>
<li><p>Ensure your V gene and allele calling is accurate, or at the very least that the contextual sequence lacks polymorphisms or errors in its very 5’.</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">stitchr</span></code> will attempt to detect and deal with single nucleotide mismatches with the stated allele, but more complex polymorphisms will result in a failure.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</section>
<section id="other-optional-arguments">
<h2>Other optional arguments<a class="headerlink" href="#other-optional-arguments" title="Permalink to this heading">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-h</span></code> - see a help menu, containing all the command line options</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-c</span></code> - specify a particular constant region gene (in the case of TRBC) or allele</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-s</span></code> - specify a species: ‘HUMAN’ is the default, see <a class="reference internal" href="inputdata.html#species-covered-label"><span class="std std-ref">Species covered</span></a> section for all options (which must be downloaded with <cite>stitchrdl</cite> or manually produced prior to use)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-aa</span></code> - provide an incomplete amino acid sequence (spanning at least the CDR3, with some padding on either side), to assess the accuracy of the stitched TCR sequence. Must be a single string,unbroken by spaces or linebreaks</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-cu</span></code> - specify the path to an alternative codon usage file, from which to generate the sequences for the non-templated residues (see the <a class="reference internal" href="inputdata.html#codon-files-label"><span class="std std-ref">Codon usage files</span></a> section)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-p</span></code> - specify a path containing gene allele preferences (see the <a class="reference internal" href="inputdata.html#preferred-allele-label"><span class="std std-ref">Preferred allele files</span></a> section)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-l</span></code> - use a different leader region to that present with the given V</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-n</span></code> - provide a name for the TCR chain, which will be included in the FASTA file header</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-3p</span></code> - provide a sequence to come immediately after the end of the constant region (e.g. a stop codon)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-5p</span></code> - provide a sequence to come immediately before the start of the L1 leader sequence (e.g. a Kozak sequence)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span></code> - define an output mode, to define which sequences get printed to the terminal</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-xg</span></code> - toggle providing additional/custom genes to be stitched into TCR transcripts in the Data/additional-genes.fasta file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-sc</span></code> - toggle skipping the constant region gene check (for genes not present in the C-region-motifs.tsv file)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-sw</span></code> - suppress warning text, which may be especially useful in conjunction with some of the alternative output modes (see the <a class="reference internal" href="#output-modes-label"><span class="std std-ref">Output modes</span></a> section)</p></li>
</ul>
<section id="output-modes">
<span id="output-modes-label"></span><h3>Output modes<a class="headerlink" href="#output-modes" title="Permalink to this heading">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">stitchr</span></code> can output the TCR sequences it generates in a number of different formats, which <a class="reference external" href="https://github.com/JamieHeather/stitchr/issues/22">may help integrate its output into certain pipelines</a>. These modes can be specified using the <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">/</span> <span class="pre">--mode</span></code> flag, using one of the following options:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">BOTH_FA</span></code></p>
<ul>
<li><p>Default option</p></li>
<li><p>Outputs a horizontal line, followed by the full, formatted, descriptive FASTA sequence of the stitched TCR, both nucleotide and translated amino acid sequence</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">NT_FA</span></code></p>
<ul>
<li><p>Outputs a horizontal line and the FASTA nucleotide sequence of the stitched TCR</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">AA_FA</span></code></p>
<ul>
<li><p>Outputs a horizontal line and the FASTA translated amino acid sequence of the stitched TCR</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">NT</span></code></p>
<ul>
<li><p>Outputs just the nucleotide sequence of the stitched TCR (no lines, no linebreaks, no FASTA header)</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">AA</span></code></p>
<ul>
<li><p>Outputs just the translated amino acid sequence of the stitched TCR (no lines, no linebreaks, no FASTA header)</p></li>
</ul>
</li>
</ul>
</section>
<section id="providing-a-partial-amino-acid-sequence">
<h3>Providing a partial amino acid sequence<a class="headerlink" href="#providing-a-partial-amino-acid-sequence" title="Permalink to this heading">¶</a></h3>
<p>If you provide a partial amino acid sequence using the <code class="docutils literal notranslate"><span class="pre">-aa</span></code> flag, <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> will perform a rudimentary pairwise alignment, just to give a quick visual assessment of the quality of the sequence generation.</p>
<p>As an example, let’s take the example of the well described A2-NLV restricted <a class="reference external" href="https://www.rcsb.org/structure/5d2n">C25 TCR from the 5D2N PDB structure</a>. We can take the amino acid sequence straight from the PDB FASTA file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="mi">5</span><span class="n">D2N</span><span class="p">:</span><span class="n">E</span><span class="o">|</span><span class="n">PDBID</span><span class="o">|</span><span class="n">CHAIN</span><span class="o">|</span><span class="n">SEQUENCE</span>
<span class="n">MGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTI</span>
<span class="n">QRTEQRDSAMYRCASSLAPGTTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVEL</span>
<span class="n">SWWVNGKEVHSGVCTDPQPLKEQPALNDSRYALSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSA</span>
<span class="n">EAWGRAD</span>
</pre></div>
</div>
<p>We can then pull out the V, J, and CDR3 information. There’s lots of ways to do this, but the easiest manual way is to find the CDR3 and then search the immediately neighbouring sequences against V/J amino acid sequences (obtainable via IMGT/GENE-DB). This gives:</p>
<p><code class="docutils literal notranslate"><span class="pre">TRBV7-6</span> <span class="pre">/</span> <span class="pre">TRBJ1-4</span> <span class="pre">/</span> <span class="pre">CASSLAPGTTNEKLFF</span></code></p>
<p>Then we can run the code like this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>stitchr<span class="w"> </span>-v<span class="w"> </span>TRBV7-6<span class="w"> </span>-j<span class="w"> </span>TRBJ1-4<span class="w"> </span>-cdr3<span class="w"> </span>CASSLAPGTTNEKLFF<span class="w"> </span>-n<span class="w"> </span>C25<span class="w"> </span>-aa<span class="w"> </span>MGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVCTDPQPLKEQPALNDSRYALSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRAD

<span class="c1"># Produces the following output</span>

&gt;nt-C25-TRBV7-6*01-TRBJ1-4*01-TRBC1*01-CASSLAPGTTNEKLFF-leader-TRBV7-6*01
ATGGGCACCAGTCTCCTATGCTGGGTGGTCCTGGGTTTCCTAGGGACAGATCACACAGGTGCTGGAGTCTCCCAGTCTCCCAGGTACAAAGTCACAAAGAGGGGACAGGATGTAGCTCTCAGGTGTGATCCAATTTCGGGTCATGTATCCCTTTATTGGTACCGACAGGCCCTGGGGCAGGGCCCAGAGTTTCTGACTTACTTCAATTATGAAGCCCAACAAGACAAATCAGGGCTGCCCAATGATCGGTTCTCTGCAGAGAGGCCTGAGGGATCCATCTCCACTCTGACGATCCAGCGCACAGAGCAGCGGGACTCGGCCATGTATCGCTGTGCCAGCAGCCTGGCCCCCGGCACCACTAATGAAAAACTGTTTTTTGGCAGTGGAACCCAGCTCTCTGTCTTGGAGGACCTGAACAAGGTGTTCCCACCCGAGGTCGCTGTGTTTGAGCCATCAGAAGCAGAGATCTCCCACACCCAAAAGGCCACACTGGTGTGCCTGGCCACAGGCTTCTTCCCCGACCACGTGGAGCTGAGCTGGTGGGTGAATGGGAAGGAGGTGCACAGTGGGGTCAGCACGGACCCGCAGCCCCTCAAGGAGCAGCCCGCCCTCAATGACTCCAGATACTGCCTGAGCAGCCGCCTGAGGGTCTCGGCCACCTTCTGGCAGAACCCCCGCAACCACTTCCGCTGTCAAGTCCAGTTCTACGGGCTCTCGGAGAATGACGAGTGGACCCAGGATAGGGCCAAACCCGTCACCCAGATCGTCAGCGCCGAGGCCTGGGGTAGAGCAGACTGTGGCTTTACCTCGGTGTCCTACCAGCAAGGGGTCCTGTCTGCCACCATCCTCTATGAGATCCTGCTAGGGAAGGCCACCCTGTATGCTGTGCTGGTCAGCGCCCTTGTGTTGATGGCCATGGTCAAGAGAAAGGATTTC

&gt;aa-C25-TRBV7-6*01-TRBJ1-4*01-TRBC1*01-CASSLAPGTTNEKLFF-leader-TRBV7-6*01
MGTSLLCWVVLGFLGTDHTGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF

MG------------------AGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQ
<span class="o">||</span><span class="w">                  </span><span class="o">||||||||||||||||||||||||||||||||||||||||</span>
MGTSLLCWVVLGFLGTDHTGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQ

GPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTT
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
GPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTT

NEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFY-PDHVELS
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||||||</span><span class="p">|</span><span class="w">  </span><span class="o">||||||</span><span class="p">|</span>
NEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGF-FPDHVELS

WWVNGKEVHSGVC-TDPQPLKEQPALNDSRYA-LSSRLRVSATFWQNPRNHFRCQVQFYG
<span class="o">||||||||||||</span><span class="w">  </span><span class="o">||||||||||||||||</span><span class="p">|</span><span class="w">  </span><span class="o">||||||||||||||||||||||||||</span><span class="p">|</span>
WWVNGKEVHSGV-STDPQPLKEQPALNDSRY-CLSSRLRVSATFWQNPRNHFRCQVQFYG

LSENDEWTQDRAKPVTQIVSAEAWGRAD--------------------------------
<span class="o">||||||||||||||||||||||||||||</span>
LSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVL

---------------

VSALVLMAMVKRKDF
</pre></div>
</div>
<p>We can see that there’s a few mismatches in the latter half of the stitched sequence, so perhaps this crystal actually used the other TRBC gene. We can swap that in:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>stitchr<span class="w"> </span>-v<span class="w"> </span>TRBV7-6<span class="w"> </span>-j<span class="w"> </span>TRBJ1-4<span class="w"> </span>-cdr3<span class="w"> </span>CASSLAPGTTNEKLFF<span class="w"> </span>-n<span class="w"> </span>C25<span class="w"> </span>-c<span class="w"> </span>TRBC2<span class="w"> </span>-aa<span class="w"> </span>MGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVCTDPQPLKEQPALNDSRYALSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRAD

<span class="c1"># Produces:</span>

&gt;nt-C25-TRBV7-6*01-TRBJ1-4*01-TRBC2*01-CASSLAPGTTNEKLFF-leader-TRBV7-6*01
ATGGGCACCAGTCTCCTATGCTGGGTGGTCCTGGGTTTCCTAGGGACAGATCACACAGGTGCTGGAGTCTCCCAGTCTCCCAGGTACAAAGTCACAAAGAGGGGACAGGATGTAGCTCTCAGGTGTGATCCAATTTCGGGTCATGTATCCCTTTATTGGTACCGACAGGCCCTGGGGCAGGGCCCAGAGTTTCTGACTTACTTCAATTATGAAGCCCAACAAGACAAATCAGGGCTGCCCAATGATCGGTTCTCTGCAGAGAGGCCTGAGGGATCCATCTCCACTCTGACGATCCAGCGCACAGAGCAGCGGGACTCGGCCATGTATCGCTGTGCCAGCAGCCTGGCCCCCGGCACCACTAATGAAAAACTGTTTTTTGGCAGTGGAACCCAGCTCTCTGTCTTGGAGGACCTGAAAAACGTGTTCCCACCCGAGGTCGCTGTGTTTGAGCCATCAGAAGCAGAGATCTCCCACACCCAAAAGGCCACACTGGTGTGCCTGGCCACAGGCTTCTACCCCGACCACGTGGAGCTGAGCTGGTGGGTGAATGGGAAGGAGGTGCACAGTGGGGTCAGCACAGACCCGCAGCCCCTCAAGGAGCAGCCCGCCCTCAATGACTCCAGATACTGCCTGAGCAGCCGCCTGAGGGTCTCGGCCACCTTCTGGCAGAACCCCCGCAACCACTTCCGCTGTCAAGTCCAGTTCTACGGGCTCTCGGAGAATGACGAGTGGACCCAGGATAGGGCCAAACCTGTCACCCAGATCGTCAGCGCCGAGGCCTGGGGTAGAGCAGACTGTGGCTTCACCTCCGAGTCTTACCAGCAAGGGGTCCTGTCTGCCACCATCCTCTATGAGATCTTGCTAGGGAAGGCCACCTTGTATGCCGTGCTGGTCAGTGCCCTCGTGCTGATGGCCATGGTCAAGAGAAAGGATTCCAGAGGC

&gt;aa-C25-TRBV7-6*01-TRBJ1-4*01-TRBC2*01-CASSLAPGTTNEKLFF-leader-TRBV7-6*01
MGTSLLCWVVLGFLGTDHTGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTTNEKLFFGSGTQLSVLEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSESYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDSRG

MG------------------AGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQ
<span class="o">||</span><span class="w">                  </span><span class="o">||||||||||||||||||||||||||||||||||||||||</span>
MGTSLLCWVVLGFLGTDHTGAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQ

GPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTT
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
GPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLAPGTT

NEKLFFGSGTQLSVLEDLNK-VFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELS
<span class="o">||||||||||||||||||</span><span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="o">||||||||||||||||||||||||||||||||||||||</span><span class="p">|</span>
NEKLFFGSGTQLSVLEDL-KNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELS

WWVNGKEVHSGVC-TDPQPLKEQPALNDSRYA-LSSRLRVSATFWQNPRNHFRCQVQFYG
<span class="o">||||||||||||</span><span class="w">  </span><span class="o">||||||||||||||||</span><span class="p">|</span><span class="w">  </span><span class="o">||||||||||||||||||||||||||</span><span class="p">|</span>
WWVNGKEVHSGV-STDPQPLKEQPALNDSRY-CLSSRLRVSATFWQNPRNHFRCQVQFYG

LSENDEWTQDRAKPVTQIVSAEAWGRAD--------------------------------
<span class="o">||||||||||||||||||||||||||||</span>
LSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSESYQQGVLSATILYEILLGKATLYAVL

-----------------

VSALVLMAMVKRKDSRG
</pre></div>
</div>
<p>This produces even more mismatches! This is an instance where the constant region used in the crystal has been altered for expression/crystallization purposes.</p>
</section>
</section>
<section id="a-note-on-cdr3-c-terminal-residues">
<h2>A note on CDR3 C-terminal residues<a class="headerlink" href="#a-note-on-cdr3-c-terminal-residues" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">stitchr</span></code> assumes that the J gene will not undergo deletion past the C-terminal residue of the CDR3 junction (which occurs approximately in the middle of the J). Thus the code looks for the appropriate residue at the end of the CDR3, which in the majority of cases will be a phenylalanine (F). However in some cases it might be something else, like a W (not uncommon in human TRAJ/mice genes) or even something more exotic like a C, L or H (which occur in certain mouse J genes). Note that most of these non-F/W residues are found in J genes with a predicted <a class="reference external" href="http://www.imgt.org/IMGTScientificChart/SequenceDescription/IMGTfunctionality.html">‘ORF’ IMGT status</a>, and thus might not contribute to functioning TCRs, but <code class="docutils literal notranslate"><span class="pre">stitchr</span></code> will still let you generate a plausible sequence using them.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">stitchr</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using <code class="docutils literal notranslate"><span class="pre">stitchr</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#gene-allele-default-behaviour">Gene/allele default behaviour</a></li>
<li class="toctree-l2"><a class="reference internal" href="#translation-sequences">Translation sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#seamless-mode">Seamless mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-optional-arguments">Other optional arguments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#output-modes">Output modes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#providing-a-partial-amino-acid-sequence">Providing a partial amino acid sequence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#a-note-on-cdr3-c-terminal-residues">A note on CDR3 C-terminal residues</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="inputdata.html"><code class="docutils literal notranslate"><span class="pre">stitchr</span></code> input data</a></li>
<li class="toctree-l1"><a class="reference internal" href="thimble.html">Thimble</a></li>
<li class="toctree-l1"><a class="reference internal" href="guistitchr.html">GUI-stitchr</a></li>
<li class="toctree-l1"><a class="reference internal" href="immunoglobulins.html">Stitching immunoglobulins</a></li>
<li class="toctree-l1"><a class="reference internal" href="referencing.html">Citing <code class="docutils literal notranslate"><span class="pre">stitchr</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="referencing.html#licensing-information">Licensing information</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Getting started</a></li>
      <li>Next: <a href="inputdata.html" title="next chapter"><code class="docutils literal notranslate"><span class="pre">stitchr</span></code> input data</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Jamie Heather.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 6.1.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/usage.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>